spec/features/course/registration_spec.rb

```diff
@@ -21,12 +21,36 @@
       expect(page).not_to have_button('.register')
     end
 
-    context 'when the user is registered in the course' do
-      let!(:course_student) { create(:course_student, course: course, user: user) }
-      scenario 'Users cannot re-register for a course' do
+    context 'when the user has requested to register for a course' do
+      let!(:requested_student) { create(:course_user, course: course, user: user) }
+
+      scenario 'user can withdraw his registration request' do
         visit course_path(course)
+        expect(page).not_to have_button(I18n.t('course.user_registrations.registration.register'))
+
+        click_link I18n.t('course.user_registrations.registration.withdraw')
+        expect(current_path).to eq(course_path(course))
+        expect(page).to have_button(I18n.t('course.user_registrations.registration.register'))
+      end
+    end
+
+    context 'when the user has been approved' do
+      let!(:approved_student) { create(:course_student, course: course, user: user) }
 
-        expect(page).not_to have_button('course.user_registrations.registration.register')
+      scenario 'user cannot re-register for or withdraw from the course' do
+        visit course_path(course)
+        expect(page).not_to have_button(I18n.t('course.user_registrations.registration.register'))
+        expect(page).not_to have_link(I18n.t('course.user_registrations.registration.withdraw'))
+      end
+    end
+
+    context 'when the user has been rejected' do
+      let!(:rejected_student) { create(:course_user, :rejected, course: course, user: user) }
+
+      scenario 'user cannot re-register for or withdraw from the course' do
```

Comment:
Does this mean we have to do backend stuff if a lecturer accidentally rejects a user?